File: .skills/openclaw-skills/skills/shifenghu/aicade-galaxy-skills/scripts/invoke_artifact.py
#!/usr/bin/env python3

import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Dict, List, Optional


def main() -> None:
    try:
        args = parse_cli_args(sys.argv[1:])
        artifact_path = require_arg(args, "--artifact")
        tool_name = require_arg(args, "--tool")
        args_file = require_arg(args, "--args-file")
        debug = is_debug_enabled(os.environ.get("AICADE_GALAXY_DEBUG"))

        artifact = json.loads(Path(artifact_path).read_text(encoding="utf-8"))
        payload = json.loads(Path(args_file).read_text(encoding="utf-8"))

        if not isinstance(payload, dict):
            raise RuntimeError("--args-file must contain a JSON object.")

        log_debug(
            debug,
            {
                "event": "invoke.input",
                "artifactPath": artifact_path,
                "tool": tool_name,
                "argsFile": args_file,
                "payload": payload,
            },
        )
        result = invoke_artifact_tool(artifact, tool_name, payload)
        log_debug(
            debug,
            {
                "event": "invoke.output",
                "artifactPath": artifact_path,
                "tool": tool_name,
                "result": result,
            },
        )
        print(json.dumps(result, ensure_ascii=False, indent=2))
        sys.exit(0 if result["ok"] else 1)
    except Exception as exc:  # noqa: BLE001
        print(str(exc), file=sys.stderr)
        sys.exit(1)


def invoke_artifact_tool(artifact: dict, tool_name: str, payload: dict) -> dict:
    tool = find_tool(artifact, tool_name)
    split = split_payload(payload)

    validate_args(tool, split["args"])
    validate_response_paths(tool, split["responsePaths"])

    url = join_base_url(artifact["baseUrl"], tool["metadata"]["path"])
    headers = {
        "Accept": "application/json",
    }
    apply_auth_header(tool, headers)

    method = tool["metadata"]["method"]
    body = None
    if method == "GET":
        url = apply_query_params(url, split["args"])
    else:
        headers["Content-Type"] = "application/json"
        body = json.dumps(split["args"]).encode("utf-8")

    request = urllib.request.Request(url, headers=headers, method=method, data=body)

    try:
        with urllib.request.urlopen(request) as response:
            status = response.status
            text = response.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as exc:
        status = exc.code
        text = exc.read().decode("utf-8", errors="replace")

    raw = parse_maybe_json(text)
    if status >= 400:
        return {
            "ok": False,
            "status": status,
            "tool": tool["name"],
            "serviceId": tool["metadata"]["serviceId"],
            "error": {
                "message": build_error_message(status, raw),
                "raw": raw,
            },
        }

    return {
        "ok": True,
        "status": status,
        "tool": tool["name"],
        "serviceId": tool["metadata"]["serviceId"],
        "data": select_response_data(raw, split["responsePaths"]),
        "raw": raw,
    }


def parse_cli_args(argv: List[str]) -> Dict[str, str]:
    if len(argv) == 0 or len(argv) % 2 != 0:
        raise RuntimeError(
            "Usage: python3 scripts/invoke_artifact.py --artifact <path> --tool <name> --args-file <path>"
        )

    result: Dict[str, str] = {}
    for index in range(0, len(argv), 2):
        key = argv[index]
        value = argv[index + 1]
        if not key.startswith("--"):
            raise RuntimeError(
                "Usage: python3 scripts/invoke_artifact.py --artifact <path> --tool <name> --args-file <path>"
            )
        result[key] = value
    return result


def require_arg(args: Dict[str, str], key: str) -> str:
    value = args.get(key)
    if not value:
        raise RuntimeError(f"Missing required argument {key}.")
    return value


def find_tool(artifact: dict, tool_name: str) -> dict:
    tools = artifact.get("tools")
    if not isinstance(tools, list):
        raise RuntimeError("Invalid artifact format.")

    for tool in tools:
        if isinstance(tool, dict) and tool.get("name") == tool_name:
            return tool

    raise RuntimeError(f'Tool "{tool_name}" was not found in artifact.')


def split_payload(payload: dict) -> dict:
    args = {key: value for key, value in payload.items() if key != "responsePaths"}
    response_paths = payload.get("responsePaths", [])
    if not isinstance(response_paths, list) or any(
        not isinstance(item, str) or len(item) == 0 for item in response_paths
    ):
        raise RuntimeError('"responsePaths" must be an array of non-empty strings.')
    return {
        "args": args,
        "responsePaths": response_paths,
    }


def validate_args(tool: dict, args: dict) -> None:
    input_schema = tool.get("inputSchema", {})
    required = input_schema.get("required", [])
    properties = input_schema.get("properties", {})

    for key in required:
        if key not in args or args[key] is None:
            raise RuntimeError(f'Missing required parameter "{key}" for tool "{tool["name"]}".')

    if input_schema.get("additionalProperties") is False:
        for key in args:
            if key not in properties:
                raise RuntimeError(f'Unknown parameter "{key}" for tool "{tool["name"]}".')

    for key, value in args.items():
        property_schema = properties.get(key)
        if isinstance(property_schema, dict):
            validate_value_type(tool["name"], key, value, property_schema)


def validate_value_type(tool_name: str, key: str, value, property_schema: dict) -> None:
    expected_type = property_schema.get("type")
    if value is None or not expected_type:
        return

    valid = (
        (expected_type == "string" and isinstance(value, str))
        or (expected_type == "number" and isinstance(value, (int, float)) and not isinstance(value, bool))
        or (expected_type == "integer" and isinstance(value, int) and not isinstance(value, bool))
        or (expected_type == "boolean" and isinstance(value, bool))
        or (expected_type == "object" and isinstance(value, dict))
        or (expected_type == "array" and isinstance(value, list))
    )

    if not valid:
        raise RuntimeError(
            f'Invalid parameter "{key}" for tool "{tool_name}": expected {expected_type}.'
        )


def validate_response_paths(tool: dict, response_paths: List[str]) -> None:
    response_fields = tool.get("metadata", {}).get("responseFields", [])
    if len(response_paths) == 0 or not isinstance(response_fields, list):
        return

    available = set(response_fields)
    for response_path in response_paths:
        if response_path not in available:
            raise RuntimeError(
                f'Unknown response path "{response_path}" for tool "{tool["name"]}".'
            )


def apply_auth_header(tool: dict, headers: Dict[str, str]) -> None:
    auth = tool.get("metadata", {}).get("authentication", {})
    env_name = auth.get("envName")
    header_name = auth.get("headerName")
    required = auth.get("required", False)
    value = os.environ.get(env_name, "") if isinstance(env_name, str) else ""

    if required and not value:
        raise RuntimeError(
            f'Missing required environment variable: {env_name} for tool "{tool["name"]}".'
        )

    if value and isinstance(header_name, str):
        headers[header_name] = value


def apply_query_params(url: str, args: dict) -> str:
    parsed = urllib.parse.urlparse(url)
    existing = urllib.parse.parse_qsl(parsed.query, keep_blank_values=True)
    extra = [(key, stringify_query_value(value)) for key, value in args.items() if value is not None]
    query = urllib.parse.urlencode(existing + extra)
    return urllib.parse.urlunparse(parsed._replace(query=query))


def stringify_query_value(value) -> str:
    if isinstance(value, (str, int, float, bool)) and not isinstance(value, dict):
        return str(value)
    return json.dumps(value, ensure_ascii=False)


def parse_maybe_json(text: str):
    if len(text) == 0:
        return None
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return text


def build_error_message(status: int, raw) -> str:
    if isinstance(raw, dict):
        if isinstance(raw.get("message"), str) and raw["message"]:
            return raw["message"]
        error = raw.get("error")
        if isinstance(error, dict) and isinstance(error.get("message"), str):
            return error["message"]

    if isinstance(raw, str) and raw:
        return raw

    return f"Request failed with HTTP {status}."


def select_response_data(data, response_paths: List[str]):
    if len(response_paths) == 0:
        return data
    if len(response_paths) == 1:
        return get_value_by_path(data, response_paths[0])

    result = {}
    for path in response_paths:
        result = merge_selected_values(result, build_selected_structure(data, tokenize_path(path)))
    return result


def get_value_by_path(data, path: str):
    return get_value_by_tokens(data, tokenize_path(path))


def build_selected_structure(data, tokens: List[str]):
    if len(tokens) == 0:
        return data

    current = tokens[0]
    rest = tokens[1:]

    if current == "[]":
        if not isinstance(data, list):
            return None
        return [build_selected_structure(item, rest) for item in data]

    if not isinstance(data, dict):
        return None

    selected_child = build_selected_structure(data.get(current), rest)
    if selected_child is None:
        return None
    return {current: selected_child}


def get_value_by_tokens(data, tokens: List[str]):
    if len(tokens) == 0:
        return data

    current = tokens[0]
    rest = tokens[1:]

    if current == "[]":
        if not isinstance(data, list):
            return None
        return [get_value_by_tokens(item, rest) for item in data]

    if not isinstance(data, dict):
        return None

    return get_value_by_tokens(data.get(current), rest)


def tokenize_path(path: str) -> List[str]:
    return [item for item in path.replace("[]", ".[]").split(".") if item]


def merge_selected_values(left, right):
    if right is None:
        return left
    if left is None:
        return right

    if isinstance(left, list) and isinstance(right, list):
        length = max(len(left), len(right))
        return [
            merge_selected_values(
                left[index] if index < len(left) else None,
                right[index] if index < len(right) else None,
            )
            for index in range(length)
        ]

    if isinstance(left, dict) and isinstance(right, dict):
        result = dict(left)
        for key, value in right.items():
            result[key] = (
                merge_selected_values(result.get(key), value)
                if key in result
                else value
            )
        return result

    return right


def join_base_url(base_url: str, path: str) -> str:
    normalized_base = base_url.rstrip("/")
    normalized_path = path if path.startswith("/") else f"/{path}"
    return f"{normalized_base}{normalized_path}"


def is_debug_enabled(value: Optional[str]) -> bool:
    return bool(value) and value.lower() == "true"


def log_debug(enabled: bool, payload: dict) -> None:
    if not enabled:
        return
    print(f"[AICADE_GALAXY_DEBUG] {json.dumps(payload, ensure_ascii=False, indent=2)}")


if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/shifenghu/aicade-galaxy-skill/scripts/invoke_artifact.mjs
#!/usr/bin/env node

import { readFile } from "node:fs/promises";
import http from "node:http";
import https from "node:https";

async function main() {
  try {
    const args = parseCliArgs(process.argv.slice(2));
    const artifactPath = requireArg(args, "--artifact");
    const toolName = requireArg(args, "--tool");
    const argsFile = requireArg(args, "--args-file");
    const debug = isDebugEnabled(process.env.AICADE_GALAXY_DEBUG);

    const artifact = JSON.parse(await readFile(artifactPath, "utf8"));
    const payload = JSON.parse(await readFile(argsFile, "utf8"));

    if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
      throw new Error("--args-file must contain a JSON object.");
    }

    logDebug(debug, {
      event: "invoke.input",
      artifactPath,
      tool: toolName,
      argsFile,
      payload,
    });

    const result = await invokeArtifactTool(artifact, toolName, payload);
    logDebug(debug, {
      event: "invoke.output",
      artifactPath,
      tool: toolName,
      result,
    });
    console.log(JSON.stringify(result, null, 2));
    process.exitCode = result.ok ? 0 : 1;
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  }
}

async function invokeArtifactTool(artifact, toolName, payload) {
  const tool = findTool(artifact, toolName);
  const split = splitPayload(payload);

  console.log("Invoking tool:", toolName, payload);
  validateArgs(tool, split.args);
  validateResponsePaths(tool, split.responsePaths);

  const url = new URL(joinBaseUrl(artifact.baseUrl, tool.metadata.path));
  const headers = {
    Accept: "application/json",
  };
  applyAuthHeader(tool, headers);

  let body;
  if (tool.metadata.method === "GET") {
    applyQueryParams(url, split.args);
  } else {
    headers["Content-Type"] = "application/json";
    body = JSON.stringify(split.args);
  }

  const response = await requestJson(url.toString(), {
    method: tool.metadata.method,
    headers,
    body,
  });

  const raw = parseMaybeJson(response.body);
  if (response.statusCode >= 400) {
    return {
      ok: false,
      status: response.statusCode,
      tool: tool.name,
      serviceId: tool.metadata.serviceId,
      error: {
        message: buildErrorMessage(response.statusCode, raw),
        raw,
      },
    };
  }

  return {
    ok: true,
    status: response.statusCode,
    tool: tool.name,
    serviceId: tool.metadata.serviceId,
    data: selectResponseData(raw, split.responsePaths),
    raw,
  };
}

function parseCliArgs(argv) {
  const result = new Map();
  for (let index = 0; index < argv.length; index += 2) {
    const key = argv[index];
    const value = argv[index + 1];
    if (!key?.startsWith("--") || value === undefined) {
      throw new Error(
        "Usage: node scripts/invoke_artifact.mjs --artifact <path> --tool <name> --args-file <path>",
      );
    }
    result.set(key, value);
  }
  return result;
}

function requireArg(args, key) {
  const value = args.get(key);
  if (!value) {
    throw new Error(`Missing required argument ${key}.`);
  }
  return value;
}

function findTool(artifact, toolName) {
  if (
    !artifact ||
    typeof artifact !== "object" ||
    !Array.isArray(artifact.tools)
  ) {
    throw new Error("Invalid artifact format.");
  }

  const tool = artifact.tools.find((item) => item?.name === toolName);
  if (!tool) {
    throw new Error(`Tool "${toolName}" was not found in artifact.`);
  }
  return tool;
}

function splitPayload(payload) {
  const { responsePaths, ...args } = payload;

  if (responsePaths === undefined) {
    return { args, responsePaths: [] };
  }

  if (
    !Array.isArray(responsePaths) ||
    responsePaths.some((item) => typeof item !== "string" || item.length === 0)
  ) {
    throw new Error('"responsePaths" must be an array of non-empty strings.');
  }

  return { args, responsePaths };
}

function validateArgs(tool, args) {
  for (const key of tool.inputSchema.required || []) {
    if (!(key in args) || args[key] === undefined || args[key] === null) {
      throw new Error(
        `Missing required parameter "${key}" for tool "${tool.name}".`,
      );
    }
  }

  if (tool.inputSchema.additionalProperties === false) {
    for (const key of Object.keys(args)) {
      if (!(key in tool.inputSchema.properties)) {
        throw new Error(`Unknown parameter "${key}" for tool "${tool.name}".`);
      }
    }
  }

  for (const [key, value] of Object.entries(args)) {
    const property = tool.inputSchema.properties[key];
    if (!property) {
      continue;
    }
    validateValueType(tool.name, key, value, property);
  }
}

function validateValueType(toolName, key, value, property) {
  if (value === undefined || value === null || !property?.type) {
    return;
  }

  const valid =
    (property.type === "string" && typeof value === "string") ||
    (property.type === "number" &&
      typeof value === "number" &&
      Number.isFinite(value)) ||
    (property.type === "integer" &&
      typeof value === "number" &&
      Number.isInteger(value)) ||
    (property.type === "boolean" && typeof value === "boolean") ||
    (property.type === "object" &&
      typeof value === "object" &&
      !Array.isArray(value)) ||
    (property.type === "array" && Array.isArray(value));

  if (!valid) {
    throw new Error(
      `Invalid parameter "${key}" for tool "${toolName}": expected ${property.type}.`,
    );
  }
}

function validateResponsePaths(tool, responsePaths) {
  if (
    responsePaths.length === 0 ||
    !Array.isArray(tool.metadata.responseFields)
  ) {
    return;
  }

  const available = new Set(tool.metadata.responseFields);
  for (const responsePath of responsePaths) {
    if (!available.has(responsePath)) {
      throw new Error(
        `Unknown response path "${responsePath}" for tool "${tool.name}".`,
      );
    }
  }
}

function applyAuthHeader(tool, headers) {
  const envName = tool.metadata?.authentication?.envName;
  const headerName = tool.metadata?.authentication?.headerName;
  const value = envName ? process.env[envName] : undefined;

  if (tool.metadata?.authentication?.required && !value) {
    throw new Error(
      `Missing required environment variable: ${envName} for tool "${tool.name}".`,
    );
  }

  if (value && headerName) {
    headers[headerName] = value;
  }
}

function applyQueryParams(url, args) {
  for (const [key, value] of Object.entries(args)) {
    if (value === undefined || value === null) {
      continue;
    }
    url.searchParams.set(key, stringifyQueryValue(value));
  }
}

function stringifyQueryValue(value) {
  if (
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean" ||
    typeof value === "bigint"
  ) {
    return String(value);
  }

  return JSON.stringify(value);
}

function requestJson(urlString, options) {
  return new Promise((resolve, reject) => {
    const url = new URL(urlString);
    const transport = url.protocol === "https:" ? https : http;
    const request = transport.request(
      url,
      {
        method: options.method,
        headers: options.headers,
      },
      (response) => {
        const chunks = [];

        response.on("data", (chunk) => {
          chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
        });

        response.on("end", () => {
          resolve({
            statusCode: response.statusCode || 0,
            body: Buffer.concat(chunks).toString("utf8"),
          });
        });
      },
    );

    request.on("error", (error) => reject(error));
    if (options.body) {
      request.write(options.body);
    }
    request.end();
  });
}

function parseMaybeJson(text) {
  if (!text) {
    return null;
  }

  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

function buildErrorMessage(status, raw) {
  if (raw && typeof raw === "object") {
    if (typeof raw.message === "string" && raw.message.length > 0) {
      return raw.message;
    }
    if (
      raw.error &&
      typeof raw.error === "object" &&
      typeof raw.error.message === "string"
    ) {
      return raw.error.message;
    }
  }

  if (typeof raw === "string" && raw.length > 0) {
    return raw;
  }

  return `Request failed with HTTP ${status}.`;
}

function selectResponseData(data, responsePaths) {
  if (responsePaths.length === 0) {
    return data;
  }

  if (responsePaths.length === 1) {
    return getValueByPath(data, responsePaths[0]);
  }

  let result = {};
  for (const path of responsePaths) {
    result = mergeSelectedValues(
      result,
      buildSelectedStructure(data, tokenizePath(path)),
    );
  }
  return result;
}

function getValueByPath(data, path) {
  return getValueByTokens(data, tokenizePath(path));
}

function buildSelectedStructure(data, tokens) {
  if (tokens.length === 0) {
    return data;
  }

  const [current, ...rest] = tokens;
  if (current === "[]") {
    if (!Array.isArray(data)) {
      return undefined;
    }
    return data.map((item) => buildSelectedStructure(item, rest));
  }

  if (data === null || typeof data !== "object") {
    return undefined;
  }

  const value = data[current];
  const selectedChild = buildSelectedStructure(value, rest);
  if (selectedChild === undefined) {
    return undefined;
  }
  return { [current]: selectedChild };
}

function getValueByTokens(data, tokens) {
  if (tokens.length === 0) {
    return data;
  }

  const [current, ...rest] = tokens;
  if (current === "[]") {
    if (!Array.isArray(data)) {
      return undefined;
    }
    return data.map((item) => getValueByTokens(item, rest));
  }

  if (data === null || typeof data !== "object") {
    return undefined;
  }

  return getValueByTokens(data[current], rest);
}

function tokenizePath(path) {
  return path.replace(/\[\]/g, ".[]").split(".").filter(Boolean);
}

function mergeSelectedValues(left, right) {
  if (right === undefined) {
    return left;
  }
  if (left === undefined) {
    return right;
  }
  if (Array.isArray(left) && Array.isArray(right)) {
    const length = Math.max(left.length, right.length);
    return Array.from({ length }, (_, index) =>
      mergeSelectedValues(left[index], right[index]),
    );
  }
  if (isPlainObject(left) && isPlainObject(right)) {
    const result = { ...left };
    for (const [key, value] of Object.entries(right)) {
      result[key] =
        key in result ? mergeSelectedValues(result[key], value) : value;
    }
    return result;
  }
  return right;
}

function isPlainObject(value) {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function joinBaseUrl(baseUrl, path) {
  const normalizedBaseUrl = String(baseUrl || "").replace(/\/+$/, "");
  const normalizedPath = String(path || "").startsWith("/")
    ? String(path)
    : `/${String(path || "")}`;
  return `${normalizedBaseUrl}${normalizedPath}`;
}

function isDebugEnabled(value) {
  return Boolean(value) && value.toLowerCase() === "true";
}

function logDebug(enabled, payload) {
  if (!enabled) {
    return;
  }

  console.log(`[AICADE_GALAXY_DEBUG] ${JSON.stringify(payload, null, 2)}`);
}

await main();
